Client/SafeBrowsing/GSafeBrowsingDatabase.swift

```diff
@@ -0,0 +1,514 @@
+                    let request = { () -> NSBatchDeleteRequest in
+                        let request: NSFetchRequest<NSFetchRequestResult> = Threat.fetchRequest()
+                        request.predicate = NSPredicate(format: "threatType == %@", response.threatType.rawValue)
+                        
+                        let deleteRequest = NSBatchDeleteRequest(fetchRequest: request)
+                        deleteRequest.resultType = .resultTypeObjectIDs
+                        return deleteRequest
+                    }()
+                    
+                    if let result = (try? self.backgroundContext.execute(request)) as? NSBatchDeleteResult {
+                        if let deletedObjects = result.result as? [NSManagedObjectID] {
+                            NSManagedObjectContext.mergeChanges(
+                                fromRemoteContextSave: [NSDeletedObjectsKey: deletedObjects],
+                                into: [self.mainContext, self.backgroundContext]
+                            )
+                        }
+                    } else {
+                        let request: NSFetchRequest<Threat> = Threat.fetchRequest()
+                        request.predicate = NSPredicate(format: "threatType == %@", response.threatType.rawValue)
+                        
+                        (try? self.backgroundContext.fetch(request))?.forEach({
+                            self.backgroundContext.delete($0)
+                        })
+                    }
+                    
+                    self.saveContext(self.backgroundContext)
+                    self.backgroundContext.reset()
+                }
+                
+            default:
+                return completion(SafeBrowsingError("Unknown Response Type"))
+            }
+            
+            let backgroundContext = { () -> NSManagedObjectContext in
+                let context = self.persistentContainer.newBackgroundContext()
+                context.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+                return context
+            }()
+            
+            backgroundContext.performAndWait { [weak self] in
+                guard let self = self else { return }
+                let request: NSFetchRequest<Threat> = Threat.fetchRequest()
+                request.predicate = NSPredicate(format: "threatType == %@", response.threatType.rawValue)
+                request.fetchLimit = 1
+                
+                let threat: Threat = (try? backgroundContext.fetch(request))?.first ?? Threat(context: backgroundContext)
+                
+                threat.threatType = response.threatType.rawValue
+                threat.state = response.newClientState
+                
+                var hashes = (threat.hashes ?? []).compactMap({ ($0 as? ThreatHash)?.hashData })
+                
+                //Hashes must be sorted
+                //hashes.sort(by: { $0.lexicographicallyPrecedes($1) })
```

Comment:
Should this line be removed?

It appears further down this function, uncommented.

Reply:
Yeah, it is sorted lower down after indexing and filtering. Just didn't want to double sort. Comment is removed now.